File: scripts/cluster/inter_connector.py
import sqlite3
import json
from typing import Dict, List, Optional, Sequence, Tuple

OPPOSITE = {"N": "S", "S": "N", "E": "W", "W": "E"}
# Convention B: vertical flipped (N means y+1; S means y-1)
DIR_DELTA = {"N": (0, 1), "E": (1, 0), "S": (0, -1), "W": (-1, 0)}


def _is_walkable(conn: sqlite3.Connection, x: int, y: int, plane: int) -> bool:
    row = conn.execute(
        "SELECT blocked, walk_mask FROM tiles WHERE x=? AND y=? AND plane=?",
        (x, y, plane),
    ).fetchone()
    if row is None:
        return False
    blocked = int(row[0]) if row[0] is not None else 1
    walk_mask = int(row[1]) if row[1] is not None else 0
    return (blocked == 0) and (walk_mask != 0)


def _get_walk_flags(conn: sqlite3.Connection, x: int, y: int, plane: int) -> Optional[Dict[str, bool]]:
    row = conn.execute(
        "SELECT walk_data FROM tiles WHERE x=? AND y=? AND plane=?",
        (x, y, plane),
    ).fetchone()
    if not row or row[0] is None:
        return None
    try:
        data = json.loads(row[0])
        return {k: bool(v) for k, v in data.items()}
    except Exception:
        return None


def _can_cross(conn: sqlite3.Connection, x: int, y: int, plane: int, d: str) -> bool:
    dx, dy = DIR_DELTA[d]
    nx, ny = x + dx, y + dy
    a = _get_walk_flags(conn, x, y, plane) or {}
    b = _get_walk_flags(conn, nx, ny, plane) or {}
    if d == "N":
        return a.get("bottom", True) and b.get("top", True)
    if d == "S":
        return a.get("top", True) and b.get("bottom", True)
    if d == "E":
        return a.get("right", True) and b.get("left", True)
    if d == "W":
        return a.get("left", True) and b.get("right", True)
    return True


def _select_chunk_pairs(conn: sqlite3.Connection,
                         planes: Optional[Sequence[int]],
                         chunk_range: Tuple[Optional[int], Optional[int], Optional[int], Optional[int]]
                         ) -> List[Tuple[int, int, int, int, int, int, str]]:
    # We iterate entrances filtered by chunk range and optional planes
    x_min, x_max, z_min, z_max = chunk_range
    conds = ["1=1"]
    params: List[int] = []
    if x_min is not None:
        conds.append("ce.chunk_x >= ?")
        params.append(int(x_min))
    if x_max is not None:
        conds.append("ce.chunk_x <= ?")
        params.append(int(x_max))
    if z_min is not None:
        conds.append("ce.chunk_z >= ?")
        params.append(int(z_min))
    if z_max is not None:
        conds.append("ce.chunk_z <= ?")
        params.append(int(z_max))
    if planes is not None and len(planes) > 0:
        placeholders = ",".join(["?"] * len(planes))
        conds.append(f"ce.plane IN ({placeholders})")
        params.extend([int(p) for p in planes])

    sql = f"""
        SELECT ce.entrance_id, ce.chunk_x, ce.chunk_z, ce.plane, ce.x, ce.y, ce.neighbor_dir
        FROM cluster_entrances ce
        WHERE {' AND '.join(conds)}
        ORDER BY ce.chunk_x, ce.chunk_z, ce.plane, ce.entrance_id
    """
    rows = conn.execute(sql, params).fetchall()
    return [(int(r[0]), int(r[1]), int(r[2]), int(r[3]), int(r[4]), int(r[5]), str(r[6])) for r in rows]


def _neighbor_chunk(chunk_x: int, chunk_z: int, d: str) -> Tuple[int, int]:
    if d == "N":
        return (chunk_x, chunk_z + 1)
    if d == "S":
        return (chunk_x, chunk_z - 1)
    if d == "E":
        return (chunk_x + 1, chunk_z)
    if d == "W":
        return (chunk_x - 1, chunk_z)
    raise ValueError("invalid dir")


def _find_opposing_entrance(conn: sqlite3.Connection, x: int, y: int, plane: int,
                            opp_chunk_x: int, opp_chunk_z: int, opp_dir: str,
                            dx: int, dy: int) -> Optional[int]:
    # Opposing entrance is at the adjacent tile across the border (x+dx, y+dy)
    row = conn.execute(
        """
        SELECT entrance_id FROM cluster_entrances
        WHERE chunk_x=? AND chunk_z=? AND plane=? AND x=? AND y=? AND neighbor_dir=?
        """,
        (opp_chunk_x, opp_chunk_z, plane, x + dx, y + dy, opp_dir),
    ).fetchone()
    return int(row[0]) if row else None


def _delete_existing_for_scope(conn: sqlite3.Connection,
                               planes: Optional[Sequence[int]],
                               chunk_range: Tuple[Optional[int], Optional[int], Optional[int], Optional[int]]) -> None:
    # Delete interconnections where entrance_from belongs to the scope
    x_min, x_max, z_min, z_max = chunk_range
    conds = ["1=1"]
    params: List[int] = []
    if x_min is not None:
        conds.append("ce.chunk_x >= ?")
        params.append(int(x_min))
    if x_max is not None:
        conds.append("ce.chunk_x <= ?")
        params.append(int(x_max))
    if z_min is not None:
        conds.append("ce.chunk_z >= ?")
        params.append(int(z_min))
    if z_max is not None:
        conds.append("ce.chunk_z <= ?")
        params.append(int(z_max))
    if planes is not None and len(planes) > 0:
        placeholders = ",".join(["?"] * len(planes))
        conds.append(f"ce.plane IN ({placeholders})")
        params.extend([int(p) for p in planes])

    sql = f"""
        DELETE FROM cluster_interconnections
        WHERE entrance_from IN (
            SELECT ce.entrance_id FROM cluster_entrances ce WHERE {' AND '.join(conds)}
        )
    """
    conn.execute(sql, params)


def run(
    *,
    conn: sqlite3.Connection,
    planes: Optional[Sequence[int]],
    chunk_range: Tuple[Optional[int], Optional[int], Optional[int], Optional[int]],
    recompute: bool,
    dry_run: bool,
    workers: int,
    store_paths: bool,
    logger,
) -> Dict[str, int]:
    # Optionally clear previous edges for scope for reproducibility
    if recompute and not dry_run:
        _delete_existing_for_scope(conn, planes, chunk_range)

    entrances = _select_chunk_pairs(conn, planes, chunk_range)
    created = 0
    examined = 0

    for entrance_id, chunk_x, chunk_z, plane, x, y, d in entrances:
        examined += 1
        opp_dir = OPPOSITE.get(d)
        if not opp_dir:
            continue
        n_cx, n_cz = _neighbor_chunk(chunk_x, chunk_z, d)

        # sanity check: both tiles walkable and directional pass-through allowed
        if not _is_walkable(conn, x, y, plane):
            continue
        dx, dy = DIR_DELTA[d]
        if not _is_walkable(conn, x + dx, y + dy, plane):
            continue
        if not _can_cross(conn, x, y, plane, d):
            continue

        opp_id = _find_opposing_entrance(conn, x, y, plane, n_cx, n_cz, opp_dir, dx, dy)
        if opp_id is None:
            continue

        if dry_run:
            created += 2  # we would insert both directions
            continue

        # Cost model: adjacent crossing cost = 1 (tile step)
        cost = 1

        # Insert bidirectional, idempotent via ON CONFLICT DO UPDATE keeping lowest cost
        conn.execute(
            """
            INSERT INTO cluster_interconnections (entrance_from, entrance_to, cost)
            VALUES (?, ?, ?)
            ON CONFLICT(entrance_from, entrance_to)
            DO UPDATE SET cost = MIN(cluster_interconnections.cost, excluded.cost)
            """,
            (entrance_id, opp_id, cost),
        )
        conn.execute(
            """
            INSERT INTO cluster_interconnections (entrance_from, entrance_to, cost)
            VALUES (?, ?, ?)
            ON CONFLICT(entrance_from, entrance_to)
            DO UPDATE SET cost = MIN(cluster_interconnections.cost, excluded.cost)
            """,
            (opp_id, entrance_id, cost),
        )
        created += 2

    logger.info(
        "inter_connector: entrances_examined=%s inter_edges_created=%s dry_run=%s",
        examined,
        created,
        dry_run,
    )

    return {"created": created, "examined": examined}

File: rust/src/commands/cluster/inter_connector.rs
use anyhow::Result;
use rusqlite::{params, Connection, OptionalExtension};
use std::collections::BTreeSet;

use super::config::Config;
use super::db::{ensure_schema, with_tx};

#[derive(Clone, Debug, Default)]
pub struct InterStats {
    pub entrances_examined: usize,
    pub inter_edges_created: usize,
}

pub fn build_inter_edges(tiles_db: &Connection, out_db: &mut Connection, cfg: &Config) -> Result<InterStats> {
    ensure_schema(out_db)?;

    let mut stats = InterStats::default();

    // Load cost for a cardinal step from meta
    let (cost_card, _cost_diag) = load_movement_costs(out_db)?;

    // Fetch entrances from out_db
    let mut entrances: Vec<(i64, i32, i32, i32, char)> = Vec::new();
    {
        let mut q = out_db.prepare(
            "SELECT entrance_id, x, y, plane, neighbor_dir FROM cluster_entrances ORDER BY plane, x, y, entrance_id",
        )?;
        let rows = q.query_map([], |row| {
            let id: i64 = row.get(0)?;
            let x: i32 = row.get(1)?;
            let y: i32 = row.get(2)?;
            let plane: i32 = row.get(3)?;
            let d: String = row.get(4)?;
            Ok((id, x, y, plane, d.chars().next().unwrap_or('?')))
        })?;
        for r in rows { entrances.push(r?); }
    }

    // Filter by cfg (planes, chunk-range) using derived chunk indices from x,y
    entrances.retain(|&(_id, x, y, plane, _)| {
        if let Some(planes) = &cfg.planes { if !planes.contains(&plane) { return false; } }
        if let Some((xmin,xmax,zmin,zmax)) = cfg.chunk_range {
            let cx = x >> 6;
            let cz = y >> 6;
            if cx < xmin || cx > xmax || cz < zmin || cz > zmax { return false; }
        }
        true
    });

    // Deterministic order already ensured

    if !cfg.dry_run {
        with_tx(out_db, |tx| {
            // Idempotence: clear prior interconnections where entrance_from belongs to scope
            if !entrances.is_empty() {
                let mut del = tx.prepare("DELETE FROM cluster_interconnections WHERE entrance_from=?1")?;
                let mut seen: BTreeSet<i64> = BTreeSet::new();
                for (eid, _x, _y, _pl, _d) in entrances.iter() {
                    if seen.insert(*eid) {
                        del.execute(params![eid])?;
                    }
                }
            }

            // Prepared statements
            let mut is_walk_stmt = tiles_db.prepare(
                "SELECT blocked, walk_mask FROM tiles WHERE x=?1 AND y=?2 AND plane=?3",
            )?;
            let mut walk_data_stmt = tiles_db.prepare(
                "SELECT walk_data FROM tiles WHERE x=?1 AND y=?2 AND plane=?3",
            )?;

            let mut opp_q = tx.prepare(
                "SELECT entrance_id FROM cluster_entrances WHERE plane=?1 AND x=?2 AND y=?3 AND neighbor_dir=?4 LIMIT 1",
            )?;

            let mut ins = tx.prepare(
                "INSERT INTO cluster_interconnections (entrance_from, entrance_to, cost)
                 VALUES (?1,?2,?3)
                 ON CONFLICT(entrance_from, entrance_to)
                 DO UPDATE SET cost = MIN(cluster_interconnections.cost, excluded.cost)",
            )?;

            for (eid, x, y, plane, d) in entrances.iter().copied() {
                stats.entrances_examined += 1;
                let (dx, dy) = dir_delta(d);
                if dx == 0 && dy == 0 { continue; }

                // Both tiles walkable and pass-through allowed across boundary
                if !is_walkable(&mut is_walk_stmt, x, y, plane)? { continue; }
                if !is_walkable(&mut is_walk_stmt, x + dx, y + dy, plane)? { continue; }
                if !can_cross(&mut walk_data_stmt, x, y, plane, d)? { continue; }

                let opp_dir = opposite(d);
                let opp_id: Option<i64> = opp_q
                    .query_row(params![plane, x + dx, y + dy, opp_dir_as_str(opp_dir)], |r| r.get(0))
                    .optional()?;
                let Some(opp_eid) = opp_id else { continue; };

                println!("Inter");
                println!("{} {} {} {} {} {}", eid, opp_eid, cost_card, opp_eid, eid, cost_card);
                // Insert both directions with cardinal cost
                ins.execute(params![eid, opp_eid, cost_card])?;
                ins.execute(params![opp_eid, eid, cost_card])?;
                stats.inter_edges_created += 2;
            }

            Ok(())
        })?;
    } else {
        // Dry-run: estimate creations based on checks
        let mut is_walk_stmt = tiles_db.prepare(
            "SELECT blocked, walk_mask FROM tiles WHERE x=?1 AND y=?2 AND plane=?3",
        )?;
        let mut walk_data_stmt = tiles_db.prepare(
            "SELECT walk_data FROM tiles WHERE x=?1 AND y=?2 AND plane=?3",
        )?;
        for (_eid, x, y, plane, d) in entrances.iter().copied() {
            stats.entrances_examined += 1;
            let (dx, dy) = dir_delta(d);
            if dx == 0 && dy == 0 { continue; }
            if !is_walkable(&mut is_walk_stmt, x, y, plane)? { continue; }
            if !is_walkable(&mut is_walk_stmt, x + dx, y + dy, plane)? { continue; }
            if !can_cross(&mut walk_data_stmt, x, y, plane, d)? { continue; }
            stats.inter_edges_created += 2;
        }
    }

    Ok(stats)
}

fn load_movement_costs(db: &Connection) -> Result<(i64,i64)> {
    let straight: String = db.query_row("SELECT value FROM meta WHERE key='movement_cost_straight'", [], |r| r.get(0))?;
    let diagonal: String = db.query_row("SELECT value FROM meta WHERE key='movement_cost_diagonal'", [], |r| r.get(0))?;
    let sc = straight.parse::<i64>().unwrap_or(1024);
    let dc = diagonal.parse::<i64>().unwrap_or(1448);
    Ok((sc, dc))
}

fn is_walkable(stmt: &mut rusqlite::Statement<'_>, x: i32, y: i32, plane: i32) -> Result<bool> {
    let row: Option<(Option<i64>, Option<i64>)> = stmt
        .query_row(params![x, y, plane], |r| Ok((r.get(0)?, r.get(1)?)))
        .optional()?;
    if let Some((blocked, walk_mask)) = row {
        let b = blocked.unwrap_or(1);
        let w = walk_mask.unwrap_or(0);
        Ok(b == 0 && w != 0)
    } else {
        Ok(false)
    }
}

fn can_cross(stmt: &mut rusqlite::Statement<'_>, x: i32, y: i32, plane: i32, d: char) -> Result<bool> {
    let (dx, dy) = dir_delta(d);
    let nx = x + dx;
    let ny = y + dy;
    let a = walk_flags(stmt, x, y, plane)?;
    let b = walk_flags(stmt, nx, ny, plane)?;
    let allow = |m: Option<bool>| m.unwrap_or(true);
    Ok(match d {
        'N' => allow(a.get("bottom").copied()) && allow(b.get("top").copied()),
        'S' => allow(a.get("top").copied()) && allow(b.get("bottom").copied()),
        'E' => allow(a.get("right").copied()) && allow(b.get("left").copied()),
        'W' => allow(a.get("left").copied()) && allow(b.get("right").copied()),
        _ => true,
    })
}

fn walk_flags(stmt: &mut rusqlite::Statement<'_>, x: i32, y: i32, plane: i32) -> Result<std::collections::HashMap<String, bool>> {
    use serde_json::Value;
    let row: Option<Option<String>> = stmt.query_row(params![x, y, plane], |r| r.get(0)).optional()?;
    let mut out = std::collections::HashMap::new();
    if let Some(Some(s)) = row {
        if let Ok(Value::Object(map)) = serde_json::from_str::<Value>(&s) {
            for (k, v) in map.into_iter() {
                out.insert(k, v.as_bool().unwrap_or(false));
            }
        }
    }
    Ok(out)
}

fn dir_delta(d: char) -> (i32,i32) {
    match d {
        'N' => (0, 1),
        'S' => (0, -1),
        'E' => (1, 0),
        'W' => (-1, 0),
        _ => (0, 0),
    }
}

fn opposite(d: char) -> char {
    match d {
        'N' => 'S',
        'S' => 'N',
        'E' => 'W',
        'W' => 'E',
        _ => '?',
    }
}

fn opp_dir_as_str(d: char) -> &'static str { match d { 'N' => "N", 'S' => "S", 'E' => "E", 'W' => "W", _ => "?" } }

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::NamedTempFile;

    #[test]
    fn creates_bidirectional_edges_with_cardinal_cost() -> Result<()> {
        let tiles_dbf = NamedTempFile::new().unwrap();
        let out_dbf = NamedTempFile::new().unwrap();
        let mut tiles = Connection::open(tiles_dbf.path())?;
        let mut out = Connection::open(out_dbf.path())?;
        crate::db::create_tables(&mut tiles)?;
        crate::db::create_tables(&mut out)?;

        // Seed chunks
        tiles.execute("INSERT INTO chunks(chunk_x, chunk_z, chunk_size, tile_count) VALUES (0,0,64,0)", [])?;
        tiles.execute("INSERT INTO chunks(chunk_x, chunk_z, chunk_size, tile_count) VALUES (1,0,64,0)", [])?;
        out.execute("INSERT INTO chunks(chunk_x, chunk_z, chunk_size, tile_count) VALUES (0,0,64,0)", [])?;
        out.execute("INSERT INTO chunks(chunk_x, chunk_z, chunk_size, tile_count) VALUES (1,0,64,0)", [])?;

        // Two adjacent walkable tiles across boundary
        tiles.execute(
            r#"INSERT INTO tiles(x,y,plane,chunk_x,chunk_z,flag,blocked,walk_mask,blocked_mask,walk_data)
                VALUES (63,0,0,0,0,0,0,1,0,'{"top":true,"bottom":true,"left":true,"right":true}')"#,
            [],
        )?;
        tiles.execute(
            r#"INSERT INTO tiles(x,y,plane,chunk_x,chunk_z,flag,blocked,walk_mask,blocked_mask,walk_data)
                VALUES (64,0,0,1,0,0,0,1,0,'{"top":true,"bottom":true,"left":true,"right":true}')"#,
            [],
        )?;

        // Entrances on each side of border
        out.execute("INSERT INTO chunk_clusters(cluster_id, chunk_x, chunk_z, plane, label, tile_count) VALUES (1,0,0,0,0,1)", [])?;
        out.execute("INSERT INTO chunk_clusters(cluster_id, chunk_x, chunk_z, plane, label, tile_count) VALUES (2,1,0,0,0,1)", [])?;
        out.execute("INSERT INTO cluster_entrances(cluster_id, x, y, plane, neighbor_dir) VALUES (1,63,0,0,'E')", [])?;
        out.execute("INSERT INTO cluster_entrances(cluster_id, x, y, plane, neighbor_dir) VALUES (2,64,0,0,'W')", [])?;

        let cfg = Config::default();
        let stats = build_inter_edges(&tiles, &mut out, &cfg)?;
        assert_eq!(stats.entrances_examined, 2);
        let cnt: i64 = out.query_row("SELECT COUNT(*) FROM cluster_interconnections", [], |r| r.get(0))?;
        assert_eq!(cnt, 2);
        let cost_card: String = out.query_row("SELECT value FROM meta WHERE key='movement_cost_straight'", [], |r| r.get(0))?;
        let expected = cost_card.parse::<i64>().unwrap_or(1024);
        let max_cost: i64 = out.query_row("SELECT MAX(cost) FROM cluster_interconnections", [], |r| r.get(0))?;
        assert_eq!(max_cost, expected);
        Ok(())
    }
}
